task/control-plane-agent/src/mgs_gimlet/host_phase2.rs

```diff
@@ -0,0 +1,285 @@
+        };
+
+        let n = gateway_messages::serialize(tx_buf, &message).unwrap_lite();
+
+        Some(UdpMetadata {
+            addr: SP_TO_MGS_MULTICAST_ADDR,
+            port: MGS_UDP_PORT,
+            size: n as u32,
+            vid: vlan_id_from_sp_port(port),
+        })
+    }
+
+    pub(crate) fn ingest_incoming_data(
+        &mut self,
+        port: SpPort,
+        hash: [u8; 32],
+        offset: u64,
+        data: &[u8],
+    ) {
+        let current = match self.current.as_mut() {
+            Some(current)
+                if hash == current.hash && offset == current.offset =>
+            {
+                current
+            }
+            _ => return,
+        };
+
+        self.buffer.clear();
+        let n = usize::min(self.buffer.capacity(), data.len());
+        self.buffer.extend_from_slice(&data[..n]).unwrap_lite();
+
+        current.state = State::Fetched;
+        current.notify_calling_task();
+        self.last_responsive_mgs = port;
+    }
+
+    pub(crate) fn get_data(
+        &self,
+        hash: [u8; 32],
+        offset: u64,
+        data: Leased<idol_runtime::W, [u8]>,
+    ) -> Result<usize, RequestError<ControlPlaneAgentError>> {
+        match self.current.as_ref() {
+            Some(current)
+                if hash == current.hash && offset == current.offset =>
+            {
+                let n = usize::min(data.len(), self.buffer.len());
+                data.write_range(0..n, &self.buffer[..n])
+                    .map_err(|()| RequestError::went_away())?;
+                Ok(n)
+            }
+            _ => Err(ControlPlaneAgentError::DataUnavailable.into()),
+        }
+    }
+}
+
+struct CurrentRequest {
+    requesting_task: TaskId,
+    requesting_task_notification_bit: u8,
+    hash: [u8; 32],
+    offset: u64,
+    state: State,
+    retry_count: u8,
+}
+
+impl CurrentRequest {
+    fn notify_calling_task(&self) {
+        sys_post(
+            self.requesting_task,
+            1 << self.requesting_task_notification_bit,
+        );
+    }
+}
+
+enum State {
+    NeedToSendFirstMgs(SpPort),
+    WaitingForFirstMgs { port: SpPort, deadline: u64 },
+    WaitingForSecondMgs { port: SpPort, deadline: u64 },
+    Fetched,
+}
+
+impl State {
+    // If we want to send a packet, returns `Some(port)` on which we want to
+    // send that packet.
+    fn port_to_send_packet(&self) -> Option<SpPort> {
+        match self {
+            State::NeedToSendFirstMgs(port) => Some(*port),
+            State::Fetched => None,
+            State::WaitingForFirstMgs { deadline, port }
+            | State::WaitingForSecondMgs { deadline, port } => {
+                if sys_get_timer().now >= *deadline {
+                    Some(*port)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    fn timer_deadline(&self) -> Option<u64> {
+        match self {
+            State::NeedToSendFirstMgs(_) => Some(sys_get_timer().now + 1),
```

Comment:
Why not `sys_get_timer().now`, to trigger an immediate send?

Reply:
Good catch; I was copy/pasting from somewhere else where the + 1 was useful. Will change.